### ISSUE TITLE
corruption_test failed: db/corruption_test.cc:325: failed: 1 == 5

### ISSUE THREAD

**Issue by GoogleCodeExporter** [chrsmithdemos/leveldb#87]
```
What steps will reproduce the problem?
1. make corruption_test
2. for((i=1;;i++)); do
       ./corruption_test
       if [[ $? -ne 0 ]]; then
           break;
       fi
   done
3.

What is the expected output? What do you see instead?

Expected output:
==== Test CorruptionTest.Recovery
expected=100..100; got=100; bad_keys=0; bad_values=0; missed=0
expected=36..36; got=36; bad_keys=0; bad_values=0; missed=64
==== Test CorruptionTest.RecoverWriteError
==== Test CorruptionTest.NewFileErrorDuringWrite
==== Test CorruptionTest.TableFile
expected=99..99; got=99; bad_keys=0; bad_values=1; missed=0
==== Test CorruptionTest.TableFileIndexData
expected=5000..9999; got=7958; bad_keys=0; bad_values=0; missed=0
==== Test CorruptionTest.MissingDescriptor
expected=1000..1000; got=1000; bad_keys=0; bad_values=0; missed=0
==== Test CorruptionTest.SequenceNumberRecovery
==== Test CorruptionTest.CorruptedDescriptor
==== Test CorruptionTest.CompactionInputError
expected=9..9; got=9; bad_keys=0; bad_values=1; missed=0
expected=10000..10000; got=10000; bad_keys=0; bad_values=0; missed=0
==== Test CorruptionTest.CompactionInputErrorParanoid
expected=9..9; got=9; bad_keys=2; bad_values=1; missed=0
==== Test CorruptionTest.UnrelatedKeys
==== PASSED 11 tests


What I see:
==== Test CorruptionTest.Recovery
expected=100..100; got=100; bad_keys=0; bad_values=0; missed=0
expected=36..36; got=36; bad_keys=0; bad_values=0; missed=64
==== Test CorruptionTest.RecoverWriteError
==== Test CorruptionTest.NewFileErrorDuringWrite
==== Test CorruptionTest.TableFile
expected=99..99; got=99; bad_keys=0; bad_values=1; missed=0
==== Test CorruptionTest.TableFileIndexData
expected=5000..9999; got=7958; bad_keys=0; bad_values=0; missed=0
==== Test CorruptionTest.MissingDescriptor
expected=1000..1000; got=1000; bad_keys=0; bad_values=0; missed=0
==== Test CorruptionTest.SequenceNumberRecovery
==== Test CorruptionTest.CorruptedDescriptor
==== Test CorruptionTest.CompactionInputError
expected=9..9; got=9; bad_keys=0; bad_values=1; missed=0
expected=10000..10000; got=10000; bad_keys=0; bad_values=0; missed=0
==== Test CorruptionTest.CompactionInputErrorParanoid
db/corruption_test.cc:325: failed: 1 == 5

What version of the product are you using? On what operating system?
1.4, git commit "85584d497e7b354853b72f450683d59fcf6b9c5c"
OS: CentOS 5.4, kernel 2.6.35.13, x64 virtual machine(VMWare)

Please provide any additional information below.


```

---

Original issue reported on code.google.com by `tfengjun` on 7 May 2012 at 6:21

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
You need to run the for loop for serveral minutes before you see the failed 
result.
If you're lucky enough, you can see it happens very soon(it once failed in the 
second run). 
```

---

Original comment by `tfengjun` on 7 May 2012 at 6:41

**Comment by GoogleCodeExporter** [comment]
```
I was not able to reproduce the bug in a real linux box, but it happened on 3 
vm instances runing CentOS 5.4, with different kernels including: 2.6.18-238, 
2.6.18-164, 2.6.35.8 and 2.6.35.13.

Maybe it's a vm issue??
```

---

Original comment by `tfengjun` on 8 May 2012 at 2:07

**Comment by GoogleCodeExporter** [comment]
```
In Ubuntu this happens on real hardware without VMs, but only on ARM: 
https://launchpad.net/ubuntu/+source/leveldb/1.10.0-1/+build/4616412
https://launchpad.net/ubuntu/+source/leveldb/1.10.0-1git1/+build/4640626

What does the "leveldb.num-files-at-level0" property mean, anyway? Is that test 
maybe too strict, and some ">= 1" might be more appropriate?
```

---

Original comment by `der.pi...@googlemail.com` on 4 Jun 2013 at 3:32

**Comment by GoogleCodeExporter** [comment]
```
Note, I re-tried the second build 
(https://launchpad.net/ubuntu/+source/leveldb/1.10.0-1git1/+build/4640626) and 
that worked the second time, so this confirms the race condition from the 
initial description, and that it is not related to a VM.
```

---

Original comment by `der.pi...@googlemail.com` on 4 Jun 2013 at 3:52

**Comment by GoogleCodeExporter** [comment]
```
der.pitti, I don't see the CorruptionTest.Recovery issue at all in the links 
you gave, just an internal compiler error in the first one. Am I missing 
something?
```

---

Original comment by `dgrogan@chromium.org` on 5 Jun 2013 at 4:51

**Comment by GoogleCodeExporter** [comment]
```
Indeed, I'm sorry. It seems that only happened in the 1git1 build, and as I 
retried the build until that "1 == 5" failure went away. It looked exactly like 
the one in the description.
```

---

Original comment by `der.pi...@googlemail.com` on 6 Jun 2013 at 8:05

**Comment by GoogleCodeExporter** [comment]
```
Apparently this is due to a race in test:

https://code.google.com/p/leveldb/issues/detail?id=182#c3
```

---

Original comment by `fullung@gmail.com` on 28 Jun 2013 at 4:04

**Comment by GoogleCodeExporter** [comment]
```
I have been able to reproduce this problem and I am working on a fix.  Thanks 
for the reports.

```

---

Original comment by `san...@google.com` on 1 Jul 2013 at 9:31

**Comment by GoogleCodeExporter** [comment]
```
Fixed in 1.13
```

---

Original comment by `dgrogan@chromium.org` on 10 Dec 2013 at 7:51
- Changed state: **Fixed**
